File: typescript/src/vault/basePoolMath.ts
import { MathSol } from '../utils/math';
import { Rounding } from './types';

export function computeAddLiquidityUnbalanced(
    currentBalances: bigint[],
    exactAmounts: bigint[],
    totalSupply: bigint,
    swapFeePercentage: bigint,
    computeInvariant: (balances: bigint[], rounding: Rounding) => bigint,
): { bptAmountOut: bigint; swapFeeAmounts: bigint[] } {
    /***********************************************************************
		//                                                                    //
		// s = totalSupply                                 (iFees - iCur)     //
		// b = tokenBalance                  bptOut = s *  --------------     //
		// bptOut = bptAmountOut                                iCur          //
		// iFees = invariantWithFeesApplied                                   //
		// iCur = currentInvariant                                            //
		// iNew = newInvariant                                                //
		***********************************************************************/

    // Determine the number of tokens in the pool.
    const numTokens = currentBalances.length;

    // Create a new array to hold the updated balances after the addition.
    const newBalances: bigint[] = new Array(numTokens);
    // Create a new array to hold the swap fee amount for each token.
    const swapFeeAmounts: bigint[] = new Array(numTokens).fill(0n);

    // Loop through each token, updating the balance with the added amount.
    for (let index = 0; index < currentBalances.length; index++) {
        newBalances[index] = currentBalances[index] + exactAmounts[index] - 1n;
    }

    // Calculate the invariant using the current balances (before the addition).
    const currentInvariant = computeInvariant(
        currentBalances,
        Rounding.ROUND_UP,
    );

    // Calculate the new invariant using the new balances (after the addition).
    const newInvariant = computeInvariant(newBalances, Rounding.ROUND_DOWN);

    // Calculate the new invariant ratio by dividing the new invariant by the old invariant.
    const invariantRatio = MathSol.divDownFixed(newInvariant, currentInvariant);

    // Loop through each token to apply fees if necessary.
    for (let index = 0; index < currentBalances.length; index++) {
        // Check if the new balance is greater than the equivalent proportional balance.
        // If so, calculate the taxable amount, rounding in favor of the protocol.
        // We round the second term down to subtract less and get a higher `taxableAmount`,
        // which charges higher swap fees. This will lower `newBalances`, which in turn lowers
        // `invariantWithFeesApplied` below.
        const proportionalTokenBalance = MathSol.mulDownFixed(
            invariantRatio,
            currentBalances[index],
        );
        if (newBalances[index] > proportionalTokenBalance) {
            const taxableAmount = newBalances[index] - proportionalTokenBalance;
            // Calculate fee amount
            swapFeeAmounts[index] = MathSol.mulUpFixed(
                taxableAmount,
                swapFeePercentage,
            );
            // Subtract the fee from the new balance.
            // We are essentially imposing swap fees on non-proportional incoming amounts.
            newBalances[index] = newBalances[index] - swapFeeAmounts[index];
        }
    }

    // Calculate the new invariant with fees applied.
    const invariantWithFeesApplied = computeInvariant(
        newBalances,
        Rounding.ROUND_DOWN,
    );

    // Calculate the amount of BPT to mint. This is done by multiplying the
    // total supply with the ratio of the change in invariant.
    // Since we multiply and divide we don't need to use FP math.
    // Round down since we're calculating BPT amount out. This is the most important result of this function,
    // equivalent to:
    // `totalSupply * (invariantWithFeesApplied / currentInvariant - 1)`

    // Then, to round `bptAmountOut` down we use `invariantWithFeesApplied` rounded down and `currentInvariant`
    // rounded up.
    // If rounding makes `invariantWithFeesApplied` smaller or equal to `currentInvariant`, this would effectively
    // be a donation. In that case we just let checked math revert for simplicity; it's not a valid use-case to
    // support at this point.
    const bptAmountOut =
        (totalSupply * (invariantWithFeesApplied - currentInvariant)) /
        currentInvariant;
    return { bptAmountOut, swapFeeAmounts };
}

export function computeAddLiquiditySingleTokenExactOut(
    currentBalances: bigint[],
    tokenInIndex: number,
    exactBptAmountOut: bigint,
    totalSupply: bigint,
    swapFeePercentage: bigint,
    computeBalance: (
        balancesLiveScaled18: bigint[],
        tokenInIndex: number,
        newSupply: bigint,
    ) => bigint,
): {
    amountInWithFee: bigint;
    swapFeeAmounts: bigint[];
} {
    // Calculate new supply after minting exactBptAmountOut
    const newSupply = exactBptAmountOut + totalSupply;
    // Calculate the initial amount of the input token needed for the desired amount of BPT out
    // "divUp" leads to a higher "newBalance," which in turn results in a larger "amountIn."
    // This leads to receiving more tokens for the same amount of BTP minted.
    const newBalance = computeBalance(
        currentBalances,
        tokenInIndex,
        MathSol.divUpFixed(newSupply, totalSupply),
    );
    const amountIn = newBalance - currentBalances[tokenInIndex];

    // Calculate the taxable amount, which is the difference
    // between the actual amount in and the non-taxable balance
    const nonTaxableBalance = MathSol.divDownFixed(
        MathSol.mulDownFixed(newSupply, currentBalances[tokenInIndex]),
        totalSupply,
    );

    const taxableAmount =
        amountIn + currentBalances[tokenInIndex] - nonTaxableBalance;

    // Calculate the swap fee based on the taxable amount and the swap fee percentage
    const fee =
        MathSol.divUpFixed(
            taxableAmount,
            MathSol.complementFixed(swapFeePercentage),
        ) - taxableAmount;

    // Create swap fees amount array and set the single fee we charge
    const swapFeeAmounts: bigint[] = new Array(currentBalances.length);
    swapFeeAmounts[tokenInIndex] = fee;

    // Return the total amount of input token needed, including the swap fee
    const amountInWithFee = amountIn + fee;
    return { amountInWithFee, swapFeeAmounts };
}

/**
 * @notice Computes the proportional amounts of tokens to be withdrawn from the pool.
 * @dev This function computes the amount of each token that will be withdrawn in exchange for burning
 * a specific amount of pool tokens (BPT). It ensures that the amounts of tokens withdrawn are proportional
 * to the current pool balances.
 *
 * Calculation: For each token, amountOut = balance * (bptAmountIn / bptTotalSupply).
 * Rounding down is used to prevent withdrawing more than the pool can afford.
 *
 * @param balances Array of current token balances in the pool.
 * @param bptTotalSupply Total supply of the pool tokens (BPT).
 * @param bptAmountIn The amount of pool tokens that will be burned.
 * @return amountsOut Array of amounts for each token to be withdrawn.
 */
export function computeProportionalAmountsOut(
    balances: bigint[],
    bptTotalSupply: bigint,
    bptAmountIn: bigint,
): bigint[] {
    /**********************************************************************************************
	// computeProportionalAmountsOut                                                             //
	// (per token)                                                                               //
	// aO = tokenAmountOut             /        bptIn         \                                  //
	// b = tokenBalance      a0 = b * | ---------------------  |                                 //
	// bptIn = bptAmountIn             \     bptTotalSupply    /                                 //
	// bpt = bptTotalSupply                                                                      //
	**********************************************************************************************/

    // Create a new array to hold the amounts of each token to be withdrawn.
    const amountsOut: bigint[] = [];
    for (let i = 0; i < balances.length; ++i) {
        // Since we multiply and divide we don't need to use FP math.
        // Round down since we're calculating amounts out.
        amountsOut.push((balances[i] * bptAmountIn) / bptTotalSupply);
    }
    return amountsOut;
}

/**
 * @notice Computes the amount of a single token to withdraw for a given amount of BPT to burn.
 * @dev It computes the output token amount for an exact input of BPT, considering current balances,
 * total supply, and swap fees.
 *
 * @param currentBalances The current token balances in the pool.
 * @param tokenOutIndex The index of the token to be withdrawn.
 * @param exactBptAmountIn The exact amount of BPT the user wants to burn.
 * @param totalSupply The total supply of BPT in the pool.
 * @param swapFeePercentage The swap fee percentage applied to the taxable amount.
 * @param computeBalance A function pointer to the balance calculation function.
 * @return amountOutWithFee The amount of the output token the user receives, accounting for swap fees.
 */
export function computeRemoveLiquiditySingleTokenExactIn(
    currentBalances: bigint[],
    tokenOutIndex: number,
    exactBptAmountIn: bigint,
    totalSupply: bigint,
    swapFeePercentage: bigint,
    computeBalance: (
        balancesLiveScaled18: bigint[],
        tokenInIndex: number,
        newSupply: bigint,
    ) => bigint,
): { amountOutWithFee: bigint; swapFeeAmounts: bigint[] } {
    // Calculate new supply accounting for burning exactBptAmountIn
    const newSupply = totalSupply - exactBptAmountIn;
    // Calculate the new balance of the output token after the BPT burn.
    // "divUp" leads to a higher "newBalance," which in turn results in a lower "amountOut."
    // This leads to giving less tokens for the same amount of BTP burned.
    const newBalance = computeBalance(
        currentBalances,
        tokenOutIndex,
        MathSol.divUpFixed(newSupply, totalSupply),
    );

    // Compute the amount to be withdrawn from the pool.
    const amountOut = currentBalances[tokenOutIndex] - newBalance;

    const newBalanceBeforeTax = MathSol.mulDivUpFixed(
        newSupply,
        currentBalances[tokenOutIndex],
        totalSupply,
    );

    // Compute the taxable amount: the difference between the new proportional and disproportional balances.
    const taxableAmount = newBalanceBeforeTax - newBalance;

    // Calculate the swap fee on the taxable amount.
    const fee = MathSol.mulUpFixed(taxableAmount, swapFeePercentage);

    // Create swap fees amount array and set the single fee we charge.
    const swapFeeAmounts = new Array(currentBalances.length);
    swapFeeAmounts[tokenOutIndex] = fee;

    // Return the net amount after subtracting the fee.
    const amountOutWithFee = amountOut - fee;

    return {
        amountOutWithFee,
        swapFeeAmounts,
    };
}

/**
 * @notice Computes the amount of pool tokens to burn to receive exact amount out.
 * @param currentBalances Current pool balances, in token registration order
 * @param tokenOutIndex Index of the token to receive in exchange for pool tokens burned
 * @param exactAmountOut Exact amount of tokens to receive
 * @param totalSupply Current total supply of the pool tokens (BPT)
 * @param swapFeePercentage The swap fee percentage applied to the taxable amount
 * @return bptAmountIn Amount of pool tokens to burn
 * @return swapFeeAmounts The amount of swap fees charged for each token
 */
export function computeRemoveLiquiditySingleTokenExactOut(
    currentBalances: bigint[],
    tokenOutIndex: number,
    exactAmountOut: bigint,
    totalSupply: bigint,
    swapFeePercentage: bigint,
    computeInvariant: (balances: bigint[], rounding: Rounding) => bigint,
): {
    bptAmountIn: bigint;
    swapFeeAmounts: bigint[];
} {
    // Determine the number of tokens in the pool.
    const numTokens = currentBalances.length;

    // Create a new array to hold the updated balances.
    const newBalances = new Array(numTokens);

    // Copy currentBalances to newBalances
    for (let index = 0; index < currentBalances.length; index++) {
        newBalances[index] = currentBalances[index] - 1n;
    }
    // Update the balance of tokenOutIndex with exactAmountOut.
    newBalances[tokenOutIndex] = newBalances[tokenOutIndex] - exactAmountOut;

    // Calculate the invariant using the current balances.
    const currentInvariant = computeInvariant(
        currentBalances,
        Rounding.ROUND_UP,
    );

    // We round invariant ratio up (see reason below).
    // This invariant ratio could be rounded up even more by rounding `currentInvariant` down. But since it only
    // affects the taxable amount and the fee calculation, whereas `currentInvariant` affects BPT in more directly,
    // we use `currentInvariant` rounded up here as well.
    const invariantRatio = MathSol.divUpFixed(
        computeInvariant(newBalances, Rounding.ROUND_UP),
        currentInvariant,
    );

    // Taxable amount is proportional to invariant ratio; a larger taxable amount rounds in the Vault's favor.
    const taxableAmount =
        MathSol.mulUpFixed(invariantRatio, currentBalances[tokenOutIndex]) -
        newBalances[tokenOutIndex];

    const fee =
        MathSol.divUpFixed(
            taxableAmount,
            MathSol.complementFixed(swapFeePercentage),
        ) - taxableAmount;

    // Update new balances array with a fee
    newBalances[tokenOutIndex] = newBalances[tokenOutIndex] - fee;

    // Calculate the new invariant with fees applied.
    const invariantWithFeesApplied = computeInvariant(
        newBalances,
        Rounding.ROUND_DOWN,
    );

    // Create swap fees amount array and set the single fee we charge
    const swapFeeAmounts = new Array(numTokens);
    swapFeeAmounts[tokenOutIndex] = fee;
    // Calculate the amount of BPT to burn. This is done by multiplying the total supply by the ratio of the
    // invariant delta to the current invariant.
    //
    // Calculating BPT amount in, so we round up. This is the most important result of this function, equivalent to:
    // `totalSupply * (1 - invariantWithFeesApplied / currentInvariant)`.
    // Then, to round `bptAmountIn` up we use `invariantWithFeesApplied` rounded down and `currentInvariant`
    // rounded up.
    //
    // Since `currentInvariant` is rounded up and `invariantWithFeesApplied` is rounded down, the difference
    // should always be positive. The checked math will revert if that is not the case.
    const bptAmountIn = MathSol.mulDivUpFixed(
        totalSupply,
        currentInvariant - invariantWithFeesApplied,
        currentInvariant,
    );

    return {
        bptAmountIn,
        swapFeeAmounts,
    };
}

File: rust/src/vault/base_pool_math.rs
//! Base pool math functions for vault operations

use crate::common::errors::PoolError;
use crate::common::maths::{
    complement_fixed, div_down_fixed, div_up_fixed, mul_div_up_fixed, mul_down_fixed, mul_up_fixed,
};
use crate::common::types::Rounding;
use num_bigint::BigInt;
use num_traits::Zero;

/// Result of add liquidity unbalanced operation
#[derive(Debug, Clone)]
pub struct AddLiquidityUnbalancedResult {
    pub bpt_amount_out: BigInt,
    pub swap_fee_amounts: Vec<BigInt>,
}

/// Result of add liquidity single token exact out operation
#[derive(Debug, Clone)]
pub struct AddLiquiditySingleTokenExactOutResult {
    pub amount_in_with_fee: BigInt,
    pub swap_fee_amounts: Vec<BigInt>,
}

/// Result of remove liquidity single token exact in operation
#[derive(Debug, Clone)]
pub struct RemoveLiquiditySingleTokenExactInResult {
    pub amount_out_with_fee: BigInt,
    pub swap_fee_amounts: Vec<BigInt>,
}

/// Result of remove liquidity single token exact out operation
#[derive(Debug, Clone)]
pub struct RemoveLiquiditySingleTokenExactOutResult {
    pub bpt_amount_in: BigInt,
    pub swap_fee_amounts: Vec<BigInt>,
}

/// Compute add liquidity for unbalanced amounts
pub fn compute_add_liquidity_unbalanced(
    current_balances: &[BigInt],
    exact_amounts: &[BigInt],
    total_supply: &BigInt,
    swap_fee_percentage: &BigInt,
    max_invariant_ratio: &BigInt,
    compute_invariant: &dyn Fn(&[BigInt], Rounding) -> Result<BigInt, PoolError>,
) -> Result<AddLiquidityUnbalancedResult, PoolError> {
    let num_tokens = current_balances.len();

    // Create new balances with added amounts
    let mut new_balances = vec![BigInt::zero(); num_tokens];
    let mut swap_fee_amounts = vec![BigInt::zero(); num_tokens];

    // Loop through each token, updating the balance with the added amount
    for index in 0..current_balances.len() {
        new_balances[index] = &current_balances[index] + &exact_amounts[index] - BigInt::from(1);
    }

    // Calculate current and new invariants
    let current_invariant = compute_invariant(current_balances, Rounding::RoundUp)?;
    let new_invariant = compute_invariant(&new_balances, Rounding::RoundDown)?;

    // Calculate invariant ratio
    let invariant_ratio = div_down_fixed(&new_invariant, &current_invariant)?;

    // Check invariant ratio bounds
    if &invariant_ratio > max_invariant_ratio {
        return Err(PoolError::MathOverflow);
    }

    // Apply fees to non-proportional amounts
    for index in 0..current_balances.len() {
        let proportional_token_balance =
            mul_down_fixed(&invariant_ratio, &current_balances[index])?;
        if new_balances[index] > proportional_token_balance {
            let taxable_amount = &new_balances[index] - &proportional_token_balance;
            let fee_amount = mul_up_fixed(&taxable_amount, swap_fee_percentage)?;
            swap_fee_amounts[index] = fee_amount.clone();
            new_balances[index] = &new_balances[index] - &fee_amount;
        }
    }

    // Calculate invariant with fees applied
    let invariant_with_fees_applied = compute_invariant(&new_balances, Rounding::RoundDown)?;

    // Calculate BPT amount out
    let bpt_amount_out =
        (total_supply * (&invariant_with_fees_applied - &current_invariant)) / &current_invariant;

    Ok(AddLiquidityUnbalancedResult {
        bpt_amount_out,
        swap_fee_amounts,
    })
}

/// Compute add liquidity for single token exact out
pub fn compute_add_liquidity_single_token_exact_out(
    current_balances: &[BigInt],
    token_in_index: usize,
    exact_bpt_amount_out: &BigInt,
    total_supply: &BigInt,
    swap_fee_percentage: &BigInt,
    max_invariant_ratio: &BigInt,
    compute_balance: &dyn Fn(&[BigInt], usize, &BigInt) -> Result<BigInt, PoolError>,
) -> Result<AddLiquiditySingleTokenExactOutResult, PoolError> {
    let new_supply = exact_bpt_amount_out + total_supply;
    let invariant_ratio = div_up_fixed(&new_supply, total_supply)?;

    // Check invariant ratio bounds
    if &invariant_ratio > max_invariant_ratio {
        return Err(PoolError::MathOverflow);
    }

    // Calculate new balance needed
    let new_balance = compute_balance(current_balances, token_in_index, &invariant_ratio)?;
    let amount_in = &new_balance - &current_balances[token_in_index];

    // Calculate non-taxable balance
    let non_taxable_balance = div_down_fixed(
        &mul_down_fixed(&new_supply, &current_balances[token_in_index])?,
        total_supply,
    )?;

    let taxable_amount = &amount_in + &current_balances[token_in_index] - &non_taxable_balance;

    // Calculate fee
    let fee =
        div_up_fixed(&taxable_amount, &complement_fixed(swap_fee_percentage)?)? - &taxable_amount;

    // Create swap fees array
    let mut swap_fee_amounts = vec![BigInt::zero(); current_balances.len()];
    swap_fee_amounts[token_in_index] = fee.clone();

    let amount_in_with_fee = &amount_in + &fee;
    Ok(AddLiquiditySingleTokenExactOutResult {
        amount_in_with_fee,
        swap_fee_amounts,
    })
}

/// Compute proportional amounts out for remove liquidity
pub fn compute_proportional_amounts_out(
    balances: &[BigInt],
    bpt_total_supply: &BigInt,
    bpt_amount_in: &BigInt,
) -> Result<Vec<BigInt>, PoolError> {
    let mut amounts_out = Vec::with_capacity(balances.len());

    for balance in balances {
        let amount_out = (balance * bpt_amount_in) / bpt_total_supply;
        amounts_out.push(amount_out);
    }

    Ok(amounts_out)
}

/// Compute remove liquidity single token exact in
pub fn compute_remove_liquidity_single_token_exact_in(
    current_balances: &[BigInt],
    token_out_index: usize,
    exact_bpt_amount_in: &BigInt,
    total_supply: &BigInt,
    swap_fee_percentage: &BigInt,
    min_invariant_ratio: &BigInt,
    compute_balance: &dyn Fn(&[BigInt], usize, &BigInt) -> Result<BigInt, PoolError>,
) -> Result<RemoveLiquiditySingleTokenExactInResult, PoolError> {
    // Calculate new supply accounting for burning exact_bpt_amount_in
    let new_supply = total_supply - exact_bpt_amount_in;

    let invariant_ratio = div_up_fixed(&new_supply, total_supply)?;

    // Check invariant ratio bounds
    if &invariant_ratio < min_invariant_ratio {
        return Err(PoolError::MathOverflow);
    }

    // Calculate the new balance of the output token after the BPT burn
    let new_balance = compute_balance(current_balances, token_out_index, &invariant_ratio)?;

    // Compute the amount to be withdrawn from the pool
    let amount_out = &current_balances[token_out_index] - &new_balance;

    let new_balance_before_tax = mul_div_up_fixed(
        &new_supply,
        &current_balances[token_out_index],
        total_supply,
    )?;

    // Compute the taxable amount: the difference between the non-taxable balance and actual withdrawal
    let taxable_amount = &new_balance_before_tax - &new_balance;

    // Calculate the swap fee on the taxable amount
    let fee = mul_up_fixed(&taxable_amount, swap_fee_percentage)?;

    // Create swap fees array
    let mut swap_fee_amounts = vec![BigInt::zero(); current_balances.len()];
    swap_fee_amounts[token_out_index] = fee.clone();

    // Return the net amount after subtracting the fee
    let amount_out_with_fee = &amount_out - &fee;
    Ok(RemoveLiquiditySingleTokenExactInResult {
        amount_out_with_fee,
        swap_fee_amounts,
    })
}

/// Compute remove liquidity single token exact out
pub fn compute_remove_liquidity_single_token_exact_out(
    current_balances: &[BigInt],
    token_out_index: usize,
    exact_amount_out: &BigInt,
    total_supply: &BigInt,
    swap_fee_percentage: &BigInt,
    min_invariant_ratio: &BigInt,
    compute_invariant: &dyn Fn(&[BigInt], Rounding) -> Result<BigInt, PoolError>,
) -> Result<RemoveLiquiditySingleTokenExactOutResult, PoolError> {
    let num_tokens = current_balances.len();

    // Create new balances array
    let mut new_balances = vec![BigInt::zero(); num_tokens];

    // Copy current_balances to new_balances
    for index in 0..current_balances.len() {
        new_balances[index] = &current_balances[index] - BigInt::from(1);
    }

    // Update the balance of token_out_index with exact_amount_out
    new_balances[token_out_index] = &new_balances[token_out_index] - exact_amount_out;

    // Calculate the invariant using the current balances
    let current_invariant = compute_invariant(current_balances, Rounding::RoundUp)?;

    let invariant_ratio = div_up_fixed(
        &compute_invariant(&new_balances, Rounding::RoundUp)?,
        &current_invariant,
    )?;

    // Check invariant ratio bounds
    if &invariant_ratio < min_invariant_ratio {
        return Err(PoolError::MathOverflow);
    }

    // Taxable amount is proportional to invariant ratio
    let taxable_amount = &mul_up_fixed(&invariant_ratio, &current_balances[token_out_index])?
        - &new_balances[token_out_index];

    let fee =
        div_up_fixed(&taxable_amount, &complement_fixed(swap_fee_percentage)?)? - &taxable_amount;

    // Update new balances array with a fee
    new_balances[token_out_index] = &new_balances[token_out_index] - &fee;

    // Calculate the new invariant with fees applied
    let invariant_with_fees_applied = compute_invariant(&new_balances, Rounding::RoundDown)?;

    // Create swap fees array
    let mut swap_fee_amounts = vec![BigInt::zero(); num_tokens];
    swap_fee_amounts[token_out_index] = fee.clone();

    // Calculate the amount of BPT to burn
    let bpt_amount_in = mul_div_up_fixed(
        total_supply,
        &(&current_invariant - &invariant_with_fees_applied),
        &current_invariant,
    )?;

    Ok(RemoveLiquiditySingleTokenExactOutResult {
        bpt_amount_in,
        swap_fee_amounts,
    })
}
